adapter/src/hermes/adapter/stdio/stdio.cc

```diff
@@ -0,0 +1,993 @@
+  int ret = -1;
+  if (hermes::adapter::IsTracked(stream)) {
+    auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+    auto existing = mdm->Find(stream);
+    if (existing.second) {
+      LOG(INFO) << "Intercept getc." << std::endl;
+      unsigned char value;
+      auto ret_size = read_internal(existing, &value, sizeof(value), stream);
+      if (ret_size == 1) {
+        ret = value;
+      }
+    } else {
+      MAP_OR_FAIL(fgetc);
+      ret = real_fgetc_(stream);
+    }
+  } else {
+    MAP_OR_FAIL(fgetc);
+    ret = real_fgetc_(stream);
+  }
+  return (ret);
+}
+
+/* NOTE: stdio.h typically implements getc() as a macro pointing to _IO_getc */
+int HERMES_DECL(_IO_getc)(FILE *stream) {
+  int ret;
+  if (hermes::adapter::IsTracked(stream)) {
+    auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+    auto existing = mdm->Find(stream);
+    if (existing.second) {
+      LOG(INFO) << "Intercept _IO_getc." << std::endl;
+      unsigned char value;
+      auto ret_size = read_internal(existing, &value, sizeof(value), stream);
+      if (ret_size == 1) {
+        ret = value;
+      }
+    } else {
+      MAP_OR_FAIL(_IO_getc);
+      ret = real__IO_getc_(stream);
+    }
+  } else {
+    MAP_OR_FAIL(_IO_getc);
+    ret = real__IO_getc_(stream);
+  }
+  return (ret);
+}
+
+/* NOTE: stdio.h typically implements putc() as a macro pointing to _IO_putc */
+int HERMES_DECL(_IO_putc)(int c, FILE *stream) {
+  int ret;
+  if (hermes::adapter::IsTracked(stream)) {
+    auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+    auto existing = mdm->Find(stream);
+    if (existing.second) {
+      LOG(INFO) << "Intercept _IO_putc char:" << (char *)&c << "." << std::endl;
+      ret = write_internal(existing, &c, 1, stream);
+    } else {
+      MAP_OR_FAIL(_IO_putc);
+      ret = real__IO_putc_(c, stream);
+    }
+  } else {
+    MAP_OR_FAIL(_IO_putc);
+    ret = real__IO_putc_(c, stream);
+  }
+  return (ret);
+}
+
+int HERMES_DECL(getw)(FILE *stream) {
+  int ret;
+  if (hermes::adapter::IsTracked(stream)) {
+    auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+    auto existing = mdm->Find(stream);
+    if (existing.second) {
+      LOG(INFO) << "Intercept getw." << std::endl;
+      unsigned char value;
+      auto ret_size = read_internal(existing, &value, sizeof(value), stream);
+      if (ret_size == 1) ret = value;
+    } else {
+      MAP_OR_FAIL(getw);
+      ret = real_getw_(stream);
+    }
+  } else {
+    MAP_OR_FAIL(getw);
+    ret = real_getw_(stream);
+  }
+  return (ret);
+}
+
+char *HERMES_DECL(fgets)(char *s, int size, FILE *stream) {
+  char *ret;
+  if (hermes::adapter::IsTracked(stream)) {
+    auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+    auto existing = mdm->Find(stream);
+    if (existing.second) {
+      LOG(INFO) << "Intercept fgets." << std::endl;
+      read_internal(existing, s, size, stream);
```

Comment:
I don't think this preserves the semantics of [fgets](https://linux.die.net/man/3/fgets):

> fgets() reads in at most one less than size characters from stream and stores them into the buffer pointed to by s. Reading stops after an EOF or a newline. If a newline is read, it is stored into the buffer. A terminating null byte (aq\0aq) is stored after the last character in the buffer.

Reply:
So i need to do the following.
Read size - 1 data
if I find \0 or \n before size - 1 i should stop and return that buffer. Correct?

Reply:
Yes, and include a null terminating byte.

Reply:
Ok fixed.